Added mdx file for Fine-tuning for Object Detection file.

Anindyadeep opened the pull request:
This PR adds the mdx file for fine-tuning ViT for object Detection.

Anindyadeep commented:
Hi @merveenoyan, sorry for adding this PR super late. I got very less bandwidth this month. But please do feel free to let me know whatever changes I need to do this PR I added, and let's close it as soon as possible. 


Some doubts I have:

- I see the quality checks are failing, but I did not see any make file in the repo, so was not able to do `make style`

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+To make this tutorial interesting, let's consider a real-world example. Consider this scenario: construction workers require the utmost safety when working in construction areas. Basic safety protocol requires wearing them helmet every time. Since there are lot of construction workers, it is hard to keep and eye on everyone everytime.
+
+But, if we can have a camera system, which can detect persons and whether the person is wearing a helmet or not in real time, that would be awesome, right?
```
```suggestion
But, if we can have a camera system that can detect persons and whether the person is wearing a helmet or not in real-time, that would be awesome, right?
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+But, if we can have a camera system, which can detect persons and whether the person is wearing a helmet or not in real time, that would be awesome, right?
+
+So, we are going to fine-tune a light weight object detection model for doing the same. Let's dive in.
```
```suggestion
So, we will fine-tune a light weight object detection model for doing the same. Let's dive in.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+### Dataset
+
+Keeping in mind for the above scenerio, we will be using the [hardhat](https://huggingface.co/datasets/hf-vision/hardhat) dataset provided by [Northeaster University China](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/7CBGOS). We will be using 🤗 `datasets` to download and load this dataset.
```
```suggestion
Keeping in mind for the above scenario, we will be using the [hardhat](https://huggingface.co/datasets/hf-vision/hardhat) dataset provided by [Northeaster University China](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/7CBGOS). We will be using `datasets` library to download and load this dataset.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+})
+```
+
+Above is a [Hugging Face DatasetDict](https://huggingface.co/docs/datasets/v2.17.1/en/package_reference/main_classes#datasets.DatasetDict), which is an efficient dict like structure containing the whole dataset in train and test splits. As you can under each split (train and test), we have `features` and `num_rows`. Under features we have the `image` which is a [Pillow Object](https://realpython. ...
````
```suggestion
Above is a [DatasetDict](https://huggingface.co/docs/datasets/v2.17.1/en/package_reference/main_classes#datasets.DatasetDict), which is an efficient dict-like structure containing the whole dataset in train and test splits. As you can see, under each split (train and test), we have `features` and `num_rows`. Under features, we have the `image`, a [Pillow Object](https://realpython.com/image-processing-with-the-python-pillow-library/), the id of the image, height and width, and objects. 
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+  'category': ['helmet', 'helmet']}}
+```
+
+As you see `objects` is an another dict containing the object ids (which are the class ids here), the area of the objects, and the bounding box co-ordinates (bbox) and the category which are the label. Here is a more detailed explaination of each of the keys and values of a data element. 
````
```suggestion
As seen, `objects` is another dict containing the object ids (which are the class ids here), the area of the objects, the bounding box coordinates (bbox), and the category, which is the label. Here is a more detailed explanation of each of the keys and values of a data element. 
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+As you see `objects` is an another dict containing the object ids (which are the class ids here), the area of the objects, and the bounding box co-ordinates (bbox) and the category which are the label. Here is a more detailed explaination of each of the keys and values of a data element. 
+
+
+1. `image`: This is a Pillow Image object that helps to look into the image directly before even loading from the path.
```
```suggestion
- `image`: This is a Pillow Image object that helps to look into the image directly before even loading from the path.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+
+1. `image`: This is a Pillow Image object that helps to look into the image directly before even loading from the path.
+2. `image_id`: Denotes which number of image is from the train file.
```
```suggestion
- `image_id`: Denotes which number of images is from the train file.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+1. `image`: This is a Pillow Image object that helps to look into the image directly before even loading from the path.
+2. `image_id`: Denotes which number of image is from the train file.
+3. `width`: The width of the image.
```
```suggestion
- `width`: The width of the image.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+1. `image`: This is a Pillow Image object that helps to look into the image directly before even loading from the path.
+2. `image_id`: Denotes which number of image is from the train file.
+3. `width`: The width of the image.
+4. `height`: The height of the image.
```
```suggestion
- `height`: The height of the image.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+2. `image_id`: Denotes which number of image is from the train file.
+3. `width`: The width of the image.
+4. `height`: The height of the image.
+5. `objects`:  Another dictionary containing information about annotation. This contains the following:
```
```suggestion
- `objects`:  Another dictionary containing information about annotation. This contains the following:
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+3. `width`: The width of the image.
+4. `height`: The height of the image.
+5. `objects`:  Another dictionary containing information about annotation. This contains the following:
+    1. `id`: A list, where length of list denotes the number of objects and value of each denotes the class index.
```
```suggestion
    - `id`: A list, where the length of the list denotes the number of objects and the value of each denotes the class index.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+4. `height`: The height of the image.
+5. `objects`:  Another dictionary containing information about annotation. This contains the following:
+    1. `id`: A list, where length of list denotes the number of objects and value of each denotes the class index.
+    2. `area`: The area of the object.
```
```suggestion
    - `area`: The area of the object.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+    1. `id`: A list, where length of list denotes the number of objects and value of each denotes the class index.
+    2. `area`: The area of the object.
+    3. `bbox`: Denotes bounding box coordinates of the object.
+    4. `category`: The class (string) of the object.
```
```suggestion
    - `category`: The class (string) of the object.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+    3. `bbox`: Denotes bounding box coordinates of the object.
+    4. `category`: The class (string) of the object.
+
+Now let's properly extract the train and test samples. For this example, we have around 5000 training samples and 1700 test samples. 
```
```suggestion
Now let's properly extract the train and test samples. We have around 5000 training samples and 1700 test samples for this example. 
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+Here is what we are going to do:
+
+1. Get the image and it's corresponding height and width.
+2. Make a draw object that can easily draw text and lines on image.
```
```suggestion
- Make a draw object that can easily draw text and lines on an image.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+3. Get the annotations dict from the sample.
+4. Iterate over it.
+5. For each, get the bounding box co-ordinates, which are x (where the bounding box starts horizontally), y (where the bounding box starts vertically), w (width of the bounding box), h (height of the bounding box).
+6. Now if the bounding box measures are normalized then scale it, else leave it.
```
```suggestion
- Now, if the bounding box measures are normalized, then scale it or leave it.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+4. Iterate over it.
+5. For each, get the bounding box co-ordinates, which are x (where the bounding box starts horizontally), y (where the bounding box starts vertically), w (width of the bounding box), h (height of the bounding box).
+6. Now if the bounding box measures are normalized then scale it, else leave it.
+7. And finally draw the rectangle and the the class category text.
```
```suggestion
- And finally, draw the rectangle and the class category text.
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+> When loading the model from the same checkpoint that you used for the preprocessing, remember to pass the label2id and id2label maps that you created earlier from the dataset’s metadata. Additionally, we specify ignore_mismatched_sizes=True to replace the existing classification head with a new one.
+
+```python
+
````
```suggestion
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+
+```
+
+Before proceeding further, login to Hugging Face, to upload your model on the fly while training. In this way you do not need to take care of handling the checkpoints and saving them somewhere. 
````
```suggestion
Before proceeding further, login to Hugging Face Hub to upload your model on the fly while training. In this way, you do not need to handle the checkpoints and save them somewhere. 
```

MKhalusova replied:
```suggestion
Before proceeding further, log in to Hugging Face to upload your model on the fly while training. In this way you do not need to take care of handling the checkpoints and saving them somewhere. 
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+plot_results(image, results)
+```
+
+### Clubbing it altogather into a function
````
```suggestion
```

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+### Clubbing it altogather into a function
+
+Now, let's club everything togather into a simple function.
```
```suggestion
Now, let's club everything together into a simple function.
```

merveenoyan reviewed (commented):
Mostly nitpicks, we can merge afterwards, thanks a lot!

MKhalusova commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+# Vision Transformers for Object Detection 
+
+In this section we are going to understand about how Object detection tasks are achieved using Vision Transformers. We will understand how we can fine-tune existing pre-trained object detection models for our specific usecase. Before getting started, check out this HuggingFace Space, where you can play around with the final output.
```
```suggestion
In this section, let's explore how to accomplish Object detection task using Vision Transformers. You will learn how to fine-tune existing pre-trained object detection models for your particular use case. Before getting started, check out this HuggingFace Space, where you can play around with the final output.
```

MKhalusova commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+## Preprocessing the dataset
+
+In this section, we will preprocess the dataset. Basically, we will apply different types of augmentations to the images, along with their corresponding bounding boxes.
```
```suggestion
Before passing the images to the `image_processor`, let's also apply different types of augmentations to the images along with their corresponding bounding boxes.
```

MKhalusova commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+
+
+```
+plot_results(image, results)
````
Add the resulting image here

Anindyadeep replied:
Added

MKhalusova commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+    plt.show()
+
+
+plot_images(test_dataset, range(6))
```
Add the output image to show the result

Anindyadeep replied:
Added

Anindyadeep commented:
Thanks for the review @merveenoyan and @MKhalusova, I will address these comments and feedback as soon as possible by tomorrow or Wednesday almost :)

merveenoyan commented:
@Anindyadeep thank you so much! we have few comments left :) after that, can you run `make quality` and add your chapter to toctree (also give yourself credit on the first chapter if you haven't)?  This way all the checks will be green 😊

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+In simple terms, augmentations are some set of random transformations like rotations, resizing etc. These are applied for the following reasons:
+
+1. To get more samples.
```
done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+There are huge number of applications around object detection. One most significant example being in the fields of autonomous driving. Where Object detection is used to detect different objects (like pedestrians, road signs, and traffic lights etc)  around the car that becomes one of the input for taking decisions.
+
+If you want to understand more around the ins-and-outs of object detection, check out our [dedicated chapter](#) on Object Detection 🤗
```
Awesome, added thanks

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+# Now use the function to plot images
+
+plot_images(train_dataset, range(9))
```
I am waiting to get this PR merged: https://huggingface.co/datasets/hf-vision/course-assets/discussions/80

Anindyadeep replied:
Done, added

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+1. To get more samples.
+2. To make the vision model more robust towards different conditions of the image.
+
+We will use the [albumentations](https://github.com/albumentations-team/albumentations) library to achieve this. If you want to dig deeper into different types of augmentations, check out the corresponding unit to learn more. Let's create some random transformations using albumentations.
```
I am not seeing any chapter done things on albumentations, (let me know if I am missing anything out here, so for now I am going to remove this part)

Anindyadeep replied:
Okay, sure

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+
+
+draw_image_from_idx(dataset=train_dataset, idx=10)
+```
````
Yes I added a PR: https://huggingface.co/datasets/hf-vision/course-assets/discussions/80, will add once merged

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+![intro](https://huggingface.co/datasets/hf-vision/course-assets/resolve/main/object_detection_wiki.png)
+
+Object detection is a computer vision task aims at identifying and localizing objects within an image or video. It involves two primary steps:
```
Done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+### Why do we need fine-tuning for object detection 🤔
+
+That is an awesome question. Before reading the answer, checkout our definitions of fine-tuning and it's difference with transfer learning [here](#).
```
Added the link

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+2. Writing repetitive model code and training them and maintaining different repositories for different use cases.
+3. Lot of experimentation and waste of resources.
+
+So, instead what we can do is, simply take a superstar pre-trained model (a model which does an awesome job in recognizing general features), and tweak or re-tune their weights (or some part of their weights) to adapt it for our use case. We believe or assume that, the pre-trained model has already learned enough to extract significant features inside an image to locate and classify objects. So,  ...
```
Done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+But, if we can have a camera system, which can detect persons and whether the person is wearing a helmet or not in real time, that would be awesome, right?
+
+So, we are going to fine-tune a light weight object detection model for doing the same. Let's dive in.
```
Done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+    3. `bbox`: Denotes bounding box coordinates of the object.
+    4. `category`: The class (string) of the object.
+
+Now let's properly extract the train and test samples. For this example, we have around 5000 training samples and 1700 test samples. 
```
Done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+test_dataset = dataset['test']
+```
+
+Now, that we know what a sample data point contains, let's start with plotting that sample. Here we are going to first draw the image and then also draw the corresponding bounding box associated.
````
Done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+Here is what we are going to do:
+
+1. Get the image and it's corresponding height and width.
```
Done

Anindyadeep commented:
> @Anindyadeep thank you so much! we have few comments left :) after that, can you run `make quality` and add your chapter to toctree (also give yourself credit on the first chapter if you haven't)? This way all the checks will be green 😊

Hey I am almost done, just to address the comments by @MKhalusova on adding images. That should be done once my asset PR in HuggingFace get's merged. Other than that, let me know if all looks good. Also where to put the credit?

merveenoyan commented:
@Anindyadeep in chapter 0 I think everyone gives credit to themselves

MKhalusova commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+Here is what we are going to do:
+
+1. Get the image and it's corresponding height and width.
```
```suggestion
1. Get the image and its corresponding height and width.
```

Anindyadeep replied:
Done

MKhalusova commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
````diff
@@ -0,0 +1,550 @@
+)
+```
+
+Before proceeding further, login to Hugging Face Hub to upload your model on the fly while training. In this way, you do not need to handle the checkpoints and save them somewhere. 
````
```suggestion
Before proceeding further, log in to Hugging Face Hub to upload your model on the fly while training. In this way, you do not need to handle the checkpoints and save them somewhere. 
```

MKhalusova reviewed (commented):
Thank you for iterating over this! I left a few minor comments.

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+### The Need to Fine-tune Models in Object Detection? 🤔
+
+That is an awesome question. Before reading the answer, check out our definitions of fine-tuning and how it differs from transfer learning [here](https://book.premai.io/state-of-open-source-ai/fine-tuning/#transfer-learning-versus-fine-tuning).
```
Done

Anindyadeep replied:
With some searches, I was't able to find any mdx reference for now. So for now I am removing this part.

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+Rather than doing all this, take a well-performing pre-trained model (a model which that does an awesome job in recognizing general features), and tweak or re-tune its weights (or some part of its weights) to adapt it for your use case. We believe or assume that the pre-trained model has already learned enough to extract significant features inside an image to locate and classify objects. So, if  ...
+
+
+
```
Done

Anindyadeep replied:
Done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+- Location of the object (XY coordinates of the bounding box)
+- Class of the object.
+
+There are a lot of of applications around object detection. One most significant examples is in the field of autonomous driving, where object detection is used to detect different objects (like pedestrians, road signs, traffic lights, etc)  around the car that become one of the inputs for taking decisions.
```
Done

Anindyadeep commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,550 @@
+
+There are a lot of of applications around object detection. One most significant examples is in the field of autonomous driving, where object detection is used to detect different objects (like pedestrians, road signs, traffic lights, etc)  around the car that become one of the inputs for taking decisions.
+
+To deepen your understanding of the ins-and-outs of object detection, check out our [dedicated chapter](https://github.com/johko/computer-vision-course/blob/main/chapters/en/Unit%206%20-%20Basic%20CV%20Tasks/object_detection.mdx) on Object Detection 🤗
```
Done

Anindyadeep commented:
Hi I addressed all the comments I guess, let me know if anything is left, and I tried to do `make quality` but I am failing this check, and upon checking it, does seem like that the error is from other chapter.

merveenoyan commented on chapters/en/Unit 3 - Vision Transformers/Vision Transformer for Objection Detection.mdx:
```diff
@@ -0,0 +1,555 @@
+
+To deepen your understanding of the ins-and-outs of object detection, check out our [dedicated chapter](/chapters/en/Unit%206%20-%20Basic%20CV%20Tasks/object_detection.mdx) on Object Detection 🤗
+
+### The Need to Fine-tune Models in Object Detection? 🤔
```
```suggestion
### The Need to Fine-tune Models in Object Detection 🤔
```

merveenoyan commented:
if you can implement the following changes, the build one will go green
https://github.com/johko/computer-vision-course/commit/8b410ff9b0210a0dbe86f540a040141f103e22c3
https://github.com/johko/computer-vision-course/commit/8b410ff9b0210a0dbe86f540a040141f103e22c3
https://github.com/johko/computer-vision-course/commit/bd5673b36b8b75a39e941a211ba10c023c349399
https://github.com/johko/computer-vision-course/commit/63cabaa80138bc3128b00163980576d58a19ce8f and we can merge

merveenoyan reviewed (approved):
Please don't merge before CI is green!

Anindyadeep commented:
Hi @merveenoyan thanks for letting me know the changes, I implemented those accordingly. Hopefully now we are good for merge :)

merveenoyan commented:
@Anindyadeep sorry apparently Maria has requested changes, we'll have to wait for her to approve this PR to merge (she's currently off)

Anindyadeep commented:
> @Anindyadeep sorry apparently Maria has requested changes, we'll have to wait for her to approve this PR to merge (she's currently off)

I see, no worries, let me know if I need to add/change anything else, would happy to do it :)

MKhalusova reviewed (approved):
Thank you for iterating over this and addressing the changes. LGTM now!

Anindyadeep commented:
> Thank you for iterating over this and addressing the changes. LGTM now!

Thanks and hopefully we can go for merge now :)
